### PR TITLE
3654 postgres lobs

### PR DESCRIPTION
**Pull request by tadgh**
* Modify Reindex to move RES_TEXT to RES_TEXT_VC if needed
* Add Test
* Add Changelog

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_3/3654-lob-problems.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+type: fix
+issue: 3654
+title: "Modify reindexing to migrate data HFJ_RES_VER data from the RES_TEXT column to the RES_TEXT_VC 
+if the resource's size falls outside of the configuration defined in DaoConfig's `getInlineResourceTextBelowSize` property."
```
Comment (fil512) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_3/3654-lob-problems.yaml:5]:
```suggestion
if the resource's size falls inside of the configuration defined in DaoConfig's `getInlineResourceTextBelowSize` property."
```

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_3/changes.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+---
+- item:
+  type: "fix"
+  title: "This version specifically modifies reindex to support moving data from the RES_TEXT to the RES_TEXT_VC column in the HFJ_RES_VER table. This is especially important
+  for PostgreSQL users, as the RES_TEXT column only has an addressable space of about 4 billion resources.
+
+Any installation that exceeds this amount of resources stored in the RES_TEXT will experience that the software hangs on attempting to store new resources.
+In order to avoid this, you should use the `DaoConfig#setInlineResourceTextBelowSize` setting, and set it to a large non-zero value.
+This will cause PostgreSQL to not store the resource text as a LOB, but instead as a VARCHAR field. By default, this field has length 4000, but you can and should update it by following the documentation [here](https://smilecdr.com/docs/fhir_storage_relational/fhir_storage_relational_module.html#resource-body-storage).
```
Comment (fil512) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_3/changes.yaml:9]: Nice!!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 3]
```diff
@@ -1466,6 +1468,26 @@ public ResourceTable updateEntity(RequestDetails theRequest, final IBaseResource
...
+	private void handleLobToTextFieldConversionForHistoricalEntities(ResourceTable theEntity) {
+		if (myDaoConfig.getInlineResourceTextBelowSize() <= 0 ) {
+			return;
+		}
+		ResourceHistoryTable currentVersionEntity = myResourceHistoryTableDao.findForIdAndVersion(theEntity.getId(), theEntity.getVersion());
+		//If the current entity is no longer in LOB territory, migrate it to text column.
+		byte[] resourceBytes = currentVersionEntity.getResource();
+		if (resourceBytes != null && resourceBytes.length > 0) {
+			String decodedResourceText = decodeResource(resourceBytes, currentVersionEntity.getEncoding());
+			if (shouldMigrateToTextField(decodedResourceText)) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1480]: Interesting!  So is the lob compressed, but the vc is not compressed??

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceHistoryTableDao.java [hunk 4]
```diff
@@ -43,6 +43,9 @@ public interface IResourceHistoryTableDao extends JpaRepository<ResourceHistoryT
 	@Query("SELECT t FROM ResourceHistoryTable t LEFT OUTER JOIN FETCH t.myProvenance WHERE t.myResourceId = :id AND t.myResourceVersion = :version")
 	ResourceHistoryTable findForIdAndVersionAndFetchProvenance(@Param("id") long theId, @Param("version") long theVersion);
 
+	@Query("SELECT t FROM ResourceHistoryTable t WHERE t.myResourceId = :id AND t.myResourceVersion = :version")
+	ResourceHistoryTable findForIdAndVersion(@Param("id") long theId, @Param("version") long theVersion);
+
 	@Query("SELECT t.myId FROM ResourceHistoryTable t WHERE t.myResourceId = :resId AND t.myResourceVersion != :dontWantVersion")
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceHistoryTableDao.java:49]: issue: remove this

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java [hunk 5]
```diff
@@ -1466,6 +1468,26 @@ public ResourceTable updateEntity(RequestDetails theRequest, final IBaseResource
 		return entity;
 	}
 
+	private void handleLobToTextFieldConversionForHistoricalEntities(ResourceTable theEntity) {
+		if (myDaoConfig.getInlineResourceTextBelowSize() <= 0 ) {
+			return;
+		}
+		ResourceHistoryTable currentVersionEntity = myResourceHistoryTableDao.findForIdAndVersion(theEntity.getId(), theEntity.getVersion());
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1475]:
```suggestion
		ResourceHistoryTable currentVersionEntity = theEntity.getCurrentVersionEntity();
```

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/delete/job/ReindexJobTest.java [hunk 6]
```diff
@@ -81,6 +89,45 @@ public void testReindexJob() throws Exception {
...
+
+		myBatchJobHelper.awaitJobCompletion(jobExecution);
+
+		//Ensure only most recent LOB is updated. First, check an older entity
+		historicalVersion =  runInTransaction(() -> myResourceHistoryTableDao.findForIdAndVersionAndFetchProvenance(obsId.getIdPartAsLong(), 10));
+		assertThat(historicalVersion.getResource(), is(notNullValue()));
+		assertThat(historicalVersion.getResourceTextVc(), is(nullValue()));
+
+		//Now, check the most recent version.
+		historicalVersion =  runInTransaction(() -> myResourceHistoryTableDao.findForIdAndVersionAndFetchProvenance(obsId.getIdPartAsLong(), 11));
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/delete/job/ReindexJobTest.java:126]: Wow this is a GREAT TEST.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/stresstest/GiantTransactionPerfTest.java [hunk 7]
```diff
@@ -354,6 +354,11 @@ public ResourceHistoryTable findForIdAndVersionAndFetchProvenance(long theId, lo
 			throw new UnsupportedOperationException();
 		}
 
+		@Override
+		public ResourceHistoryTable findForIdAndVersion(long theId, long theVersion) {
+			throw new UnsupportedOperationException();
+		}
+
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/stresstest/GiantTransactionPerfTest.java:361]: issue: remove this

**Review by fil512 — CHANGES_REQUESTED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceHistoryTableDao.java (hunk 4 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceHistoryTableDao.java:49]: done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java (hunk 5 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirDao.java:1475]: done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java [hunk 8]
```diff
@@ -1395,11 +1395,31 @@ public void reindex(T theResource, ResourceTable theEntity) {
...
 	}
+	private boolean shouldMigrateToTextField(String resourceString) {
+		return resourceString.length() < getConfig().getInlineResourceTextBelowSize();
+	}
+	private void handleLobToTextFieldConversionIfNecessary(ResourceTable theEntity) {
+		if (myDaoConfig.getInlineResourceTextBelowSize() <= 0 ) {
+			return;
+		}
+		ResourceHistoryTable currentVersionEntity = theEntity.getCurrentVersionEntity();
+		//If the current entity is no longer in LOB territory, migrate it to text column.
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java:1412]: if (currentVersionEntity == null) {
  return;
}

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/stresstest/GiantTransactionPerfTest.java (hunk 7 above)
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/stresstest/GiantTransactionPerfTest.java:361]: it's still here.  Did you forget to push?

**Review by fil512 — APPROVED** [review]

**Pull request merged**